src/net_processing.cpp

```diff
@@ -2094,6 +2107,7 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
                     return true;
                 } else if (status == READ_STATUS_FAILED) {
                     // Duplicate txindexes, the block is now in-flight, so just request it
+                    // NOTE: This is the one place two full block requests can be outstanding
```

Comment:
OK, so why not just check fPeerWasFirstRequest and MarkBlockAsNotInFlight otherwise?